ttax/riemannian.py

```diff
@@ -0,0 +1,113 @@
+import functools
+from typing import List
+import jax
+import jax.numpy as jnp
+
+from ttax.base_class import TT
+from ttax.base_class import TTMatrix
+from ttax.base_class import TTTensOrMat
+from ttax.ops import tt_vmap
+from ttax.decompositions import orthogonalize
+
+
+@tt_vmap
+def tangent_to_deltas(tangent_element: TTTensOrMat) -> List[jnp.ndarray]:
+  """Convert an element of the tangent space to deltas representation.
+  Tangent space elements (outputs of ttax.project) look like:
+    dP1 V2 ... Vd + U1 dP2 V3 ... Vd + ... + U1 ... Ud-1 dPd.
+  This function takes as input an element of the tangent space and converts
+  it to the list of deltas [dP1, ..., dPd].
+  Args:
+      tangent_element: `TT` or `TTMatrix` that is a result of ttax.project.
+  Returns:
+      A list of delta-cores.
+  """
+  # TODO: project on
+  ndim = tangent_element.ndim
+  tt_ranks = tangent_element.tt_ranks
+  deltas = [None] * ndim
+  for i in range(1, ndim - 1):
+    if int(tt_ranks[i] / 2) != float(tt_ranks[i]) / 2:
+      raise ValueError('tangent_element argument is supposed to be a '
+                       'projection, but its ranks are not even.')
+  for i in range(1, ndim - 1):
+    r1, r2 = tt_ranks[i], tt_ranks[i + 1]
+    curr_core = tangent_element.tt_cores[i]
+    deltas[i] = curr_core[int(r1 / 2):, ..., :int(r2 / 2)]
+  deltas[0] = tangent_element.tt_cores[0][..., :int(tt_ranks[1] / 2)]
+  deltas[ndim - 1] = tangent_element.tt_cores[ndim - 1][int(tt_ranks[-2] / 2):]
+  return deltas
+
+
+def _deltas_tt_vmap(_deltas_to_tangent):
+  """Like tt_vmap, but tailored to be used with deltas_to_tangent."""
+  @functools.wraps(_deltas_to_tangent)
+  def vectorized_deltas_to_tangent(deltas, tt):
+    if tt.is_tt_matrix:
+      num_batch_dims = len(deltas[0].shape) - 4
+    else:
+      num_batch_dims = len(deltas[0].shape) - 3
+    if num_batch_dims == 0:
+      return _deltas_to_tangent(deltas, tt)
+    else:
+      # Vmap everything num_batch_dims times.
+      vmapped = _deltas_to_tangent
+      for _ in range(num_batch_dims):
+        vmapped = jax.vmap(vmapped, in_axes=(0, None))
+      return vmapped(deltas, tt)
+  return vectorized_deltas_to_tangent
+
+
+@_deltas_tt_vmap
+def deltas_to_tangent(deltas: List[jnp.ndarray],
+                      tt: TTTensOrMat) -> TTTensOrMat:
+  """Converts deltas representation of tangent space vector to TT object.
+  Takes as input a list of [dP1, ..., dPd] and returns
+    dP1 V2 ... Vd + U1 dP2 V3 ... Vd + ... + U1 ... Ud-1 dPd.
+  This function is hard to use correctly because deltas should abey the
+  so called gauge conditions. If the don't, the function will silently return
+  incorrect result. This is why this function is not imported in __init__.
```

Comment:
This function is hard to use correctly because deltas should obey the
so called gauge conditions. If they don't, the function will silently return
incorrect result. That is why this function is not imported in __init__.

Reply:
Sorry, I didn't get: what's the difference?

Reply:
Just lil grammar fixes.

Reply:
Got it, thanks, fixed.